lib/estella/parser.rb

```diff
@@ -18,7 +18,10 @@ def field(name, opts = {})
       using = opts[:using] || name
       analysis = opts[:analysis] & @model.default_analysis_fields.keys
       opts[:fields] ||= Hash[analysis.zip(@model.default_analysis_fields.values_at(*analysis))] if analysis
-
+      include_raw = opts.delete(:include_raw)
+      if include_raw
+        opts[:fields][:raw] = { type: "keyword" }
```

Comment:
@joeyAghion I added this , so that it can be used like so: https://github.com/artsy/gravity/blob/16fa6e04ec10c6d1276c1473785afd79a5ee74ef/app/models/domain/sale.rb#L127 (note `include_raw`).

This creates a mapping for the `name` field like the following:

```
:name=>
    {:type=>"text",
     :fields=>
      {:default=>{:type=>"text", :analyzer=>"default_analyzer"},
       :snowball=>{:type=>"text", :analyzer=>"snowball_analyzer"},
       :shingle=>{:type=>"text", :analyzer=>"shingle_analyzer"},
       :ngram=>
        {:type=>"text",
         :analyzer=>"ngram_analyzer",
         :search_analyzer=>"search_analyzer"},
       :raw=>{:type=>"keyword"}}},
```

Notice there's a `raw` field with a `keyword` type.

We need this in order to support _both_ use cases on a name field. For instance, we want to be able to text searching on it, and so need the `text` type with the analysis setup. But, we also want to be able to filter/sort/aggregate on the `name` field (which doesn't work with a `text` type).

We then need to switch references that do sorting on filtering or aggregation for 'multi' fields such as this, by using `name.raw`, `genes.raw`, etc.

Some examples:
https://github.com/artsy/gravity/blob/16fa6e04ec10c6d1276c1473785afd79a5ee74ef/app/api/v1/related_endpoint.rb#L242

https://github.com/artsy/gravity/blob/16fa6e04ec10c6d1276c1473785afd79a5ee74ef/app/models/search/queries/artwork_filtered_query.rb#L184

Reply:
Interesting. Is this `raw` thing documented somewhere? One thing I encountered is that treating all strings as _both_ `text` and `keyword` is the new default, so I'm wondering if it's completely necessary:

https://www.elastic.co/blog/strings-are-dead-long-live-strings

Reply:
(discussed a bit in standup)- that's the default mapping applied when a field is encountered in a document and it's not in the mapping.

That will hopefully make adding fields easier (currently needing to create the mapping manually before the deploy), but shouldn't affect our mapping here.